src/theory/bv/bvgauss.cpp

```diff
@@ -0,0 +1,715 @@
+    }
+  }
+
+  bool ispart = false;
+  for (size_t i = 0; i < nrows; ++i)
+  {
+    size_t pcol = i;
+    while (pcol < ncols && reslhs[i][pcol] == 0) ++pcol;
+    if (pcol >= ncols)
+    {
+      resrhs[i] = resrhs[i].euclidianDivideRemainder(prime);
+      if (resrhs[i] != 0)
+      {
+        /* no solution */
+        return BVGaussElim::Result::NONE;
+      }
+      continue;
+    }
+    for (size_t j = i; j < ncols; ++j)
+    {
+      if (reslhs[i][j] >= prime || reslhs[i][j] <= -prime)
+      {
+        reslhs[i][j] = reslhs[i][j].euclidianDivideRemainder(prime);
+      }
+      if (j > pcol && reslhs[i][j] != 0)
+      {
+        ispart = true;
+      }
+    }
+  }
+
+  if (ispart) { return BVGaussElim::Result::PARTIAL; }
+
+  return BVGaussElim::Result::UNIQUE;
+}
+
+BVGaussElim::Result BVGaussElim::gaussElimRewriteForUrem(
+    const vector<Node>& equations,
+    unordered_map<Node, Node, NodeHashFunction>& res)
+{
+  Assert(res.empty());
+
+  Node prime;
+  Integer iprime;
+  unordered_map<Node, vector<Integer>, NodeHashFunction> vars;
+  size_t neqs = equations.size();
+  vector<Integer> resrhs, rhs;
+  vector<vector<Integer>> reslhs,
+      lhs = vector<vector<Integer>>(neqs, vector<Integer>());
+
+  res = unordered_map<Node, Node, NodeHashFunction>();
+
+  for (size_t i = 0; i < neqs; ++i)
+  {
+    Node eq = equations[i];
+    Assert(eq.getKind() == kind::EQUAL);
+    Node urem, eqrhs;
+
+    if (eq[0].getKind() == kind::BITVECTOR_UREM)
+    {
+      urem = eq[0];
+      Assert(is_bv_const(eq[1]));
+      eqrhs = eq[1];
+    }
+    else
+    {
+      Assert(eq[1].getKind() == kind::BITVECTOR_UREM);
+      urem = eq[1];
+      Assert(is_bv_const(eq[0]));
+      eqrhs = eq[0];
+    }
+    if (getMinBwExpr(Rewriter::rewrite(urem[0])) == 0)
+    {
+      Trace("bv-gauss-elim")
+          << "Minimum required bit-width exceeds given bit-width, "
+             "will not apply Gaussian Elimination."
+          << endl;
+      return BVGaussElim::Result::INVALID;
+    }
+    rhs.push_back(get_bv_const_value(eqrhs));
+
+    Assert(is_bv_const(urem[1]));
+    Assert(i == 0 || get_bv_const_value(urem[1]) == iprime);
+    if (i == 0)
+    {
+      prime = urem[1];
+      iprime = get_bv_const_value(prime);
+    }
+
+    unordered_map<Node, Integer, NodeHashFunction> tmp;
+    stack<Node> stack;
+    stack.push(urem[0]);
+    while (!stack.empty())
+    {
+      Node n = stack.top();
+      stack.pop();
+
+      /* Subtract from rhs if const */
+      if (is_bv_const(n))
+      {
+        Integer val = get_bv_const_value(n);
+        if (val > 0) rhs.back() -= val;
+        continue;
+      }
+
+      /* Split into matrix columns */
+      Kind k = n.getKind();
+      if (k == kind::BITVECTOR_PLUS)
+      {
+        for (const Node& nn : n) { stack.push(nn); }
+      }
+      else if (k == kind::BITVECTOR_MULT)
+      {
+        Node n0, n1;
+        /* Flatten mult expression. */
+        n = RewriteRule<FlattenAssocCommut>::run<true>(n);
```

Comment:
It seems like FlattenAssocCommut also flattens `BITVECTOR_PLUS`, so you could just apply it to the whole equation, which would simplify the code a bit.

Reply:
I don't see why it would simplify the code, the handling of `BITVECTOR_PLUS` would be the same. That's why I decided to only flatten where necessary (here: multiplications).

Reply:
You could just have two nested for loops, one iterating over the addition, one over the multiplications. You are currently implicitly flattening the additions anyway.

Reply:
That does not necessarily simplify things because of corner cases (which require that you either again use a stack or introduce lots of code duplication). I'll keep it as it is.